graphical_console: Add mouse related functions

peixiu opened the pull request:
Include mouse button down/up/press func,
mouse scroll up/down func,
and mouse move func.

id: 1724060, 1724061
Signed-off-by: Peixiu Hou <phou@redhat.com>

peixiu commented:
@luckyh please help to review it when you free, thanks a lot~

peixiu commented:
@luckyh Hi xuhan, I updated the code, could you help to review it?
Thanks a lot~

peixiu commented:
Hi @luckyh , I updated the code, please help to review it when you free~
Thanks a lot~

peixiu commented:
Updated the code, tested on win2019, all passed.

PASS 1-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.vioinput_mice.device_mouse.q35
PASS 2-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.vioinput_mice.device_tablet.q35

peixiu commented:
@luckyh please help to review this patch~
Thanks~

luckyh commented on virttest/graphical_console.py:
```diff
@@ -31,22 +34,76 @@ def get_keymap(keymap_file):
     def __init__(self, vm, logfile=None):
         self._vm = vm
+        self.width, self.height = self.screen_size()
+        self.pointer_pos = (int(self.width/2), int(self.height/2))
```
Can't assume the pointer always be at the centre of the screen, for example, when you reattach the console with a newly created object, the pointer may be away of the centre.

peixiu replied:
I do a pointer position initialization here, if users create a new graphical console object, the default pointer position will set to this console screen central, it will be changed along with pointer move, users will always get the latest pointer position in the same console, but if they create a new object again, the pointer position will also be re-initiated in this new console.

luckyh replied:
Let me explain about that, supposes that you have created an console object and moved the pointer to somewhere, for example the top-right, the screen would look like the following (`^` is presented as the pointer):
```
+-------+
|     ^ |
|       |
|       |
+-------+
```
then if you close the console connection and recreate a new one, by the current approach it would look like:
```
  +-------+
+ | - - + |
  |   ^   |
| |     | |
  +-------+
+ - - - +
```
the box drawn by solid line is the screen seen by the current console, and the one drawn by dash line is the screen seen by the previous, they do not coincide, and that means you have a new coordinate system now, then the question is which one should you choose to get the position of the pointer?

peixiu replied:
as we discussed, will use a class attribute to save pointer position, will get pointer position dynamically through this class attribute. Thanks a lot~

peixiu replied:
Updated the codes, new added a class method _set_pointer_pos() to change the class attribute POINTER_POS, because instance method cannot change it.
Thanks~

luckyh commented on virttest/graphical_console.py:
```diff
@@ -96,6 +153,104 @@ def _hold_time(self, hold_time):
+                time.sleep(interval)
+        self._scroll_event("up")
+
+    def pointer_move(self, pos, motion=None, mtype='abs'):
```
Suggest `mtype='abs'` -> `abs=True`

peixiu replied:
thanks, will updated it~

luckyh commented on virttest/graphical_console.py:
```diff
@@ -229,6 +415,95 @@ def _key_event(self, down, keys):
+        elif event == "down":
+            self._btn_event(self.BTN_DOWN, self.SCROLL_BACKWARD)
+        else:
+            raise ConsoleNotSupportedError("qmp console doesn't support"
```
Seems it won't happen so you can remove the `else` branch.

peixiu replied:
Considering may be need support scroll_left or scroll_right event except scroll_up/down in the future, so add this else branch here, qmp console doesn't support them now.

luckyh commented on virttest/graphical_console.py:
```diff
@@ -229,6 +415,95 @@ def _key_event(self, down, keys):
+        Returns a tuple of 2 integers
+        """
+
+        image_file = "/tmp/screendump"
```
Should use the tmp dir provided by VT
```python
from virttest import data_dir

data_dir.get_tmp_dir()
```

peixiu replied:
ok, will update, thanks

luckyh commented on virttest/graphical_console.py:
```diff
@@ -229,6 +415,95 @@ def _key_event(self, down, keys):
+        Returns a tuple of 2 integers
+        """
+
+        image_file = "/tmp/screendump"
```
Have to ensure that the filename has not been used.

peixiu replied:
ok, will update it, thanks

luckyh commented on virttest/graphical_console.py:
```diff
@@ -229,6 +415,95 @@ def _key_event(self, down, keys):
+        image_file = "/tmp/screendump"
+        self._vm.monitor.screendump(image_file)
+        img = Image.open(image_file)
+        return img.size
```
Instead of directly manipulating the image, it is better to introduce a function (i.e. `image_size`) to the `ppm_utils` module and call it at here
```python
def image_size(filename):
    img = Image.open(image_file)
    return img.size
```

peixiu replied:
thanks, will update~

peixiu commented on virttest/graphical_console.py:
```diff
@@ -96,6 +153,104 @@ def _hold_time(self, hold_time):
+        btnstroke = self._reverse(btnstroke)
+        self.btn_release(btnstroke, interval)
+
+    def scroll_down(self, count=1, interval=0):
```
yes, will update, thanks~

peixiu commented on virttest/graphical_console.py:
```diff
@@ -31,22 +34,76 @@ def get_keymap(keymap_file):
     def __init__(self, vm, logfile=None):
         self._vm = vm
+        self.width, self.height = self.screen_size()
+        self.pointer_pos = (int(self.width/2), int(self.height/2))
```
thanks, will update~

peixiu commented on virttest/graphical_console.py:
```diff
@@ -96,6 +153,104 @@ def _hold_time(self, hold_time):
+
+        count -= 1
+        for num in range(count):
+            self._scroll_event("down")
```
ok, will update it, thanks~

peixiu commented:
Hi @luckyh I updated the code, for screendump file name, I changed it to screendump + 4 random num+str characters as suffix.
when called screen_size(), all screendump_* file are save in new generated tem dir,  we can cover it if file name is same with before, we catch screen size need to use the latest screen dump file, so I do not add if file is exists codes here. Please help to review it again,  thanks a lot~

Best Regards~
Peixiu

luckyh commented on virttest/graphical_console.py:
```diff
@@ -96,6 +147,104 @@ def _hold_time(self, hold_time):
+                time.sleep(interval)
+        self._scroll_event(self.SCROLL_FORWARD)
+
+    def pointer_move(self, pos, motion=None, _abs=True):
```
Please use the full spelling `absolute` instead of `_abs`.

peixiu replied:
ok, will update~

luckyh commented on virttest/graphical_console.py:
```diff
@@ -229,6 +409,90 @@ def _key_event(self, down, keys):
+                      Does not support scroll horizontal left/right currently.
+        """
+
+        self._btn_event(self.BTN_DOWN, event)
```
typo?

peixiu replied:
For qmp_console, send btn event with qmp command as:
{'execute': 'input-send-event', 'arguments': {'events': [{'type': 'btn', 'data': {'down': True, 'button': 'left'}}]}, 'id': 'W3uMGZHZ'}
send scroll event is same with send btn event, just only has True option, none False option, like:
{'execute': 'input-send-event', 'arguments': {'events': [{'type': 'btn', 'data': {'down': True, 'button': 'wheel-up'}}]}, 'id': 'Qohmxaei'}
{'execute': 'input-send-event', 'arguments': {'events': [{'type': 'btn', 'data': {'down': True, 'button': 'wheel-down'}}]}, 'id': 'VqEnRfs1'}
Due to they have same structure, and I want to separate the _btn_event() and _scroll_event away,  so I call self._btn_event() directly for _scroll_event(),  just I should not use "event" param, it should define scroll action(wheel-up or wheel-down),  "event" is not so clear here, I want change it to "scroll", do you think so that?

luckyh replied:
Ok, thanks for the clarification! so I think we'd better 'leave a comment for that.

Btw, doesn't it need to be followed with an up event?

peixiu replied:
updated the code, and yes, qemu doesn't need to do up action for wheel-up and wheel-down, only for btn(left, right, middle,side, extra) need do up action followed down action.

luckyh commented on virttest/graphical_console.py:
```diff
@@ -229,6 +409,90 @@ def _key_event(self, down, keys):
+        """
+
+        image = data_dir.get_tmp_dir() + "/screendump_" + \
+            utils_misc.generate_random_string(4)
```
Suggest:
```python
tmp_dir = os.path.join(data_dir.get_tmp_dir(), "graphic_console_%s" % self._vm.name)
if not os.path.exists(tmp_dir):
    os.mkdir(tmp_dir)
image = os.path.join(tmp_dir, "screendump")

```

peixiu replied:
Thanks a lot, will update it~

peixiu replied:
updated~

peixiu commented:
Tested with patch https://github.com/autotest/tp-qemu/pull/1902 together:
 (1/2) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (41.74 s)
 (2/2) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.vioinput_mice.device_tablet.q35: PASS (41.84 s)

@luckyh please help to review it again, thanks a lot~

peixiu commented:
@luckyh please help to review this patch again, thanks a lot~

luckyh commented on virttest/graphical_console.py:
```diff
@@ -31,6 +34,45 @@ def get_keymap(keymap_file):
+            decimal_y += float("0." + dy_float)
+            if decimal_x >= 1:
+                x += (-1 if "-" in str(dx_int) else 1)
+                decimal_x = float("0." + (str(decimal_x).split(".")[1]))
```
Looks too complicated, is there some reason to have to use this handmade signed floor? shall we use `round` (四舍五入) instead? then the code would be more clear.
```python
def motion(src, dst):
    start_x, start_y = src
    end_x, end_y = dst
    nr_samples = (duration * rate) if duration != 0 else 1
    dx = (end_x - start_x) / nr_samples
    dy = (end_y - start_y) / nr_samples
    x, y = start_x, start_y
    interval = (duration / nr_samples) if duration != 0 else 0
    while True:
        x += dx
        y += dy
        time.sleep(interval)
        if (abs(x) >= abs(end_x)) or (abs(y) >= abs(end_y)):
            break
        yield (round(x), round(y))
    yield (end_x, end_y)

peixiu replied:
use round also is ok, thanks, has updated~

luckyh commented on virttest/graphical_console.py:
```diff
@@ -31,6 +34,45 @@ def get_keymap(keymap_file):
+            if i == num - 1:
+                x, y = x2, y2
+            yield (x, y)
+            time.sleep(duration / rate)
```
not correct, please reference the above comment.

peixiu replied:
updated~

peixiu commented on virttest/graphical_console.py:
```diff
@@ -45,8 +87,19 @@ class BaseConsole(object):
+    SCROLL_FORWARD = 1
+    SCROLL_BACKWARD = -1
+
+    POINTER_POS = None
```
that's ok, thanks, will update it~

peixiu commented on virttest/graphical_console.py:
```diff
@@ -96,6 +149,115 @@ def _hold_time(self, hold_time):
+        else:
+            self._motion_event(pos, absolute)
+
+    def screen_size(self):
```
ok, thanks, will update~

peixiu commented on virttest/graphical_console.py:
```diff
@@ -45,8 +87,19 @@ class BaseConsole(object):
+
     def __init__(self, vm, logfile=None):
         self._vm = vm
+        self._width, self._height = self.screen_size()
```
thanks, will update~

peixiu commented on virttest/graphical_console.py:
```diff
@@ -96,6 +149,115 @@ def _hold_time(self, hold_time):
+        """
+
+        if motion:
+            #for _pos in motion(self._pointer_pos, pos):
```
ok, will update~

peixiu commented on virttest/graphical_console.py:
```diff
@@ -96,6 +149,115 @@ def _hold_time(self, hold_time):
+
+        raise NotImplementedError
+
+    def get_pointer_pos(self):
```
ok, thanks, will update~

peixiu commented on virttest/graphical_console.py:
```diff
@@ -96,6 +149,115 @@ def _hold_time(self, hold_time):
+        raise NotImplementedError
+
+    @classmethod
+    def _set_pointer_pos(cls, pos):
```
hi xuhan, thanks for review, I added this class method due to there are 2 situation need to use this method.
1. we need initiate the pointer_pos in center of screen, but screen_size() is a instance method, we cannot change the class attribute value through a instance method, so if use assignment directly, we need define the class attribute under the screen_size() method, it looks not beautiful for me; currently the first place used this class method in line 102. 
2. we also need change the class attribute POINTER_POS when the pointer moved(in line 469) to track the pointer place, also hit the similar situation, we cannot assignment new value to the class attribute value in a instance. method.
So I used a class method to complete it here~

luckyh replied:
I mean you can set the value by `QMPConsole.POINTER_POS = (x, y)`, but yes it is not as clear as having a method, for reading. In fact, my true concern is that not all types of console would require `POINTER_POS` had to be an class attribute, then it is not good to implement this as a class method in the base class. However, I didn't figure out yet that how would different console manage their own pointer and display, so let's keep the approach until we find a better one, thanks.

peixiu replied:
ok, thanks~

peixiu commented on virttest/graphical_console.py:
```diff
@@ -229,6 +422,95 @@ def _key_event(self, down, keys):
+        Returns a tuple of 2 integers.
+        """
+
+        return self.POINTER_POS
```
updated~

peixiu commented:
Updated the code and tested with latest code: 
(1/2) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (50.40 s)
 (2/2) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.vioinput_mice.device_tablet.q35: PASS (48.70 s)

@luckyh Please help to review it again, thanks a lot~

peixiu commented on virttest/graphical_console.py:
```diff
@@ -96,6 +143,116 @@ def _hold_time(self, hold_time):
+        raise NotImplementedError
+
+    @property
+    def get_pointer_pos(self):
```
Updated, thanks~

peixiu commented:
Tested with latest code:
 (1/2) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (50.14 s)
 (2/2) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.vioinput_mice.device_tablet.q35: PASS (50.14 s)

luckyh commented:
Thanks all.
